debug/connection-channel.js

```diff
@@ -0,0 +1,80 @@
+/**
+ * @license
+ * Copyright (c) 2018 Google Inc. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * Code distributed by Google as part of this project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+'use strict';
+
+import WebSocket from '../platform/ws-web.js';
+import initDebug from '../devtools/src/init-debug.js';
+
+class ConnectionChannel {
+  constructor() {
+    this.debouncedMessages = [];
+    this.debouncing = false;
+    this.ready = new Promise((resolve, reject) => {
+      this._makeReady = resolve;
+    });
+  }
+
+  send(message) {
+    this.debouncedMessages.push(message);
+    if (!this.debouncing) {
+      this.debouncing = true;
+      setTimeout(() => {
+        this._flush(this.debouncedMessages);
+        this.debouncedMessages = [];
+        this.debouncing = false;
+      }, 100);
+    }
+  }
+}
+
+class ChromeExtensionChannel extends ConnectionChannel {
+  constructor() {
+    super();
+    this._makeReady(); // TODO: Consider readiness if connecting via extension.
+  }
+
+  _flush(messages) {
+    document.dispatchEvent(new CustomEvent('arcs-debug', {detail: messages}));
+  }
+}
+
+class WebSocketChannel extends ConnectionChannel {
+  constructor() {
+    super();
+    this.server = new WebSocket.Server({port: 8787});
+    this.server.on('connection', ws => {
+      this.socket = ws;
+      this.socket.on('message', msg => {
+        if (msg === 'init') {
+          initDebug();
+          this._makeReady();
+        }
+      });
+    });
+  }
+
+  _flush(messages) {
+    if (this.socket) {
+      this.socket.send(JSON.stringify(messages));
+    }
+  }
+}
+
+let _instance = null;
+export default {
+  get: () => {
+    if (!_instance) {
+      _instance = typeof window === 'object'
```

Comment:
Trying to figure out if you're in node or web is something that should happen at the platform level. eg. ChromeExtensionChannel is somethihng-web.js WebSocketChannel is something-node.js.

Reply:
Moved the implementations in the end, as this means I don't need to have platform switch for websocket, as it's only used in the node implementation of the node devtools connection.